### PR TITLE
Refactoring train.py, removing OpenCV, adding training results to Tensborboard, bug fixes

### PR DESCRIPTION
**Pull request by yalaudah**
I think moving forward, we'll use smaller PRs. But here are the changes in this one:

Fixes issue #236 that involves rewriting a big portion of `train.py` such that:
- All the tensorboard event handlers are organized in tensorboard_handlers.py and only called in train.py to log training and validation results in Tensorboard
- The code logs the same results for training and validation. Also, it adds the class IoU score as well.
- All single-use functions (e.g. _select_max, _tensor_to_numpy, _select_pred_and_mask) are lambda functions now
- The code is organized into more meaningful "chunks".. e.g. all the optimizer-related code should be together if possible, same thing for logging, configuration, loaders, tensorboard, ..etc.

In addition: 
- Fixed a visualization bug where the seismic images where not normalized correctly. This solves Issue #217.
- Fixed a visualization bug where the predictions where not masked where the input image was padded. This improves the ability to visually inspect and evaluate the results. This solves Issue #230. 
- Fixes a potential issue where Tensorboard can crash when a large training batchsize is used. Now the number of images visualized in Tensorboard from every batch has an upper limit.
- Completely removed OpenCV as a dependency from the DeepSeismic Repo. It was only used in a small part of the code where it wasn't really necessary, and OpenCV is a huge library. 
- Fixes Issue #218 where the epoch number for the images in Tensorboard was always logged as 1 (therefore, not allowing use to see the epoch number of the different results in Tensorboard. 
- Removes the `HorovodLRScheduler` class since its no longer used
- Removes toolz.take from Debug mode, and uses PyTorch's native Subset() dataset class
- Changes default patch size for the HRNet model to 256
- In addition to several other minor changes

### CODE REVIEW

File: environment/docker/apex/dockerfile [hunk 1]
```diff
@@ -10,7 +10,7 @@ RUN git clone https://github.com/NVIDIA/apex && \
     cd apex && \
     pip install -v --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" ./
 
-RUN pip install toolz pytorch-ignite torchvision pandas opencv-python fire tensorboardx scikit-learn yacs
+RUN pip install toolz pytorch-ignite torchvision pandas fire tensorboardx scikit-learn yacs
```
Comment (maxkazmsft) [inline environment/docker/apex/dockerfile:13]: Do we need to modify apex as a part of this PR?

File: experiments/interpretation/dutchf3_patch/distributed/train.py [hunk 2]
```diff
@@ -137,7 +136,7 @@ def run(*options, cfg=None, local_rank=0, debug=False):
             PadIfNeeded(
                 min_height=config.TRAIN.PATCH_SIZE,
                 min_width=config.TRAIN.PATCH_SIZE,
-                border_mode=cv2.BORDER_CONSTANT,
+                border_mode=0,
```
Comment (maxkazmsft) [inline experiments/interpretation/dutchf3_patch/distributed/train.py:139]: My suggestion would be to to define constants to mean something and place them into __init__.py, for example inside cv_lib. Then import like "from cv_lib import OPENCV_BORDER_CONSTANT" and set "border_mode=OPENCV_BORDER_CONSTANT". Then you know what the constant means - "0" isn't very descriptive.

File: experiments/interpretation/dutchf3_patch/distributed/train.py [hunk 3]
```diff
@@ -147,7 +146,7 @@ def run(*options, cfg=None, local_rank=0, debug=False):
             PadIfNeeded(
                 min_height=config.TRAIN.AUGMENTATIONS.PAD.HEIGHT,
                 min_width=config.TRAIN.AUGMENTATIONS.PAD.WIDTH,
-                border_mode=cv2.BORDER_CONSTANT,
+                border_mode=0,
```
Comment (maxkazmsft) [inline experiments/interpretation/dutchf3_patch/distributed/train.py:149]: Same comment as above

File: experiments/interpretation/dutchf3_patch/local/train.py [hunk 4]
```diff
@@ -128,7 +107,7 @@ def run(*options, cfg=None, debug=False):
             PadIfNeeded(
                 min_height=config.TRAIN.AUGMENTATIONS.PAD.HEIGHT,
                 min_width=config.TRAIN.AUGMENTATIONS.PAD.WIDTH,
-                border_mode=cv2.BORDER_CONSTANT,
+                border_mode=0,
```
Comment (maxkazmsft) [inline experiments/interpretation/dutchf3_patch/local/train.py:110]: same comment as above

File: experiments/interpretation/dutchf3_patch/local/train.py [hunk 5]
```diff
@@ -161,6 +140,11 @@ def run(*options, cfg=None, debug=False):
     )
     logger.info(val_set)
 
+    if debug:
+        logger.info("Running in Debug Mode")
+        train_set = data.Subset(train_set, list(range(config.TRAIN.BATCH_SIZE_PER_GPU)))
```
Comment (maxkazmsft) [inline experiments/interpretation/dutchf3_patch/local/train.py:145]: we should check that this works somehow... I guess the build VM will take care of this? Fix is https://github.com/microsoft/seismic-deeplearning/issues/238 so if it breaks replace with toolz.take again

File: experiments/interpretation/penobscot/local/test.py [hunk 6]
```diff
@@ -139,7 +138,7 @@ def run(*options, cfg=None, debug=False):
             PadIfNeeded(
                 min_height=config.TRAIN.PATCH_SIZE,
                 min_width=config.TRAIN.PATCH_SIZE,
-                border_mode=cv2.BORDER_CONSTANT,
+                border_mode=0,
```
Comment (maxkazmsft) [inline experiments/interpretation/penobscot/local/test.py:141]: same comment as above

File: experiments/interpretation/penobscot/local/test.py [hunk 7]
```diff
@@ -150,7 +149,7 @@ def run(*options, cfg=None, debug=False):
             PadIfNeeded(
                 min_height=config.TRAIN.AUGMENTATIONS.PAD.HEIGHT,
                 min_width=config.TRAIN.AUGMENTATIONS.PAD.WIDTH,
-                border_mode=cv2.BORDER_CONSTANT,
```
Comment (maxkazmsft) [inline experiments/interpretation/penobscot/local/test.py:153]: same comment as above

File: experiments/interpretation/penobscot/local/train.py [hunk 8]
```diff
@@ -127,7 +126,7 @@ def run(*options, cfg=None, debug=False):
             PadIfNeeded(
                 min_height=config.TRAIN.PATCH_SIZE,
                 min_width=config.TRAIN.PATCH_SIZE,
-                border_mode=cv2.BORDER_CONSTANT,
+                border_mode=0,
```
Comment (maxkazmsft) [inline experiments/interpretation/penobscot/local/train.py:129]: same comment as above

File: experiments/interpretation/penobscot/local/train.py [hunk 9]
```diff
@@ -138,7 +137,7 @@ def run(*options, cfg=None, debug=False):
             PadIfNeeded(
                 min_height=config.TRAIN.AUGMENTATIONS.PAD.HEIGHT,
                 min_width=config.TRAIN.AUGMENTATIONS.PAD.WIDTH,
-                border_mode=cv2.BORDER_CONSTANT,
+                border_mode=0,
```
Comment (maxkazmsft) [inline experiments/interpretation/penobscot/local/train.py:140]: same comment as above

File: experiments/interpretation/dutchf3_patch/local/train.py (hunk 5 above)
Comment (yalaudah) [reply experiments/interpretation/dutchf3_patch/local/train.py:145]: Yeah agreed. If this doesn't work, we should add this as a separate issue, and fix it in a future PR

File: experiments/interpretation/dutchf3_patch/distributed/train.py (hunk 2 above)
Comment (yalaudah) [reply experiments/interpretation/dutchf3_patch/distributed/train.py:139]: That makes sense, I'll add this to `__init__.py`

File: environment/docker/apex/dockerfile (hunk 1 above)
Comment (yalaudah) [reply environment/docker/apex/dockerfile:13]: are we using apex? I'm not sure, but i figured opencv would be removed from all the repo, so there would be no need to install it

Comment (maxkazmsft) [reply environment/docker/apex/dockerfile:13]: ok - doesn't do anything - please feel free to leave

File: experiments/interpretation/dutchf3_patch/distributed/train.py (hunk 2 above)
Comment (yalaudah) [reply experiments/interpretation/dutchf3_patch/distributed/train.py]: So it turned out I can't import it from `__init__.py`, but I found a better solution to add it the YACS config file. One  less import to worry about :)

**Comment by yalaudah** [comment]
sdes/azp run microsoft.ComponentGovernance (seismic-deeplearning)

**Comment by yalaudah** [comment]
Closing this since its replaced by #256 .
